docs/modules/ROOT/pages/gsn.adoc

```diff
@@ -0,0 +1,93 @@
+
+One of RelayHub's jobs is to act as a, well, _hub_ for all relayers: they will advertise their services on this contract, and your users will query it to find the relayer that best suits their purposes. This is out of scope for this guide however, and is not something you need to worry about when writing a recipient contract. If you want to learn more about sending transactions via relayers, head to our [GSNProvider guide]().
+
+The other key task RelayHub carries out is the actual _relaying of transactions_, the sole purpose behind this whole system. Instead of calling a function in your contract directly, your users will request a relayer to do it for them, who will then execute RelayHub's relayCall function. RelayHub will verify that the transaction is legitimate (protecting both users and recipients from dishonest relayers), and then call into your contract as originally requested by your user. This requires your recipient trusting RelayHub to do the right thing, but since it is a smart contract, this is as simple as reading its source code!
+
+== Receiving a relayed call
+
+We've mentioned how it is RelayHub, not your user, that actually ends up calling a function in your contract. We will refer to this as the _relayed call_. OpenZeppelin Contracts includes a number of utilities to make receiving relayed calls as easy as developing a regular Solidiy contract, without needing to worry about the low level details.
+
+The first step to writing a recipient is to inherit from our GSNRecipient contract. If you're also inheriting from other OpenZeppelin contracts, such as ERC20 or Crowdsale, this will work just fine: by adding GSNRecipient all of your token or crowdsale functions will also be GSN-callable.
+
+```solidity
+import "@openzeppelin/contracts/GSN/GSNRecipient.sol";
+
+contract MyContract is GSNRecipient, ... {
+
+}
+```
+
+=== msg.sender and msg.data
+
+There's only one extra care you should take when working with GSN recipient contracts: _you must never use `msg.sender` or `msg.data` directly_. On relayed calls, `msg.sender` will be `RelayHub` instead of your user! This doesn't mean however you won't be able to retrieve your users' addresses: `GSNRecipient` provides two functions, `_msgSender()` and `_msgData()`, which are drop-in replacements for `msg.sender` and `msg.data` while taking care of the low-level details. As long as you use these two functions instead of the original getters, you're good to go!
```

Comment:
"There's only one extra care you should take when working with GSN recipient contracts" could be
"There's only one extra detail you need to take care of when working with GSN recipient contracts"

"As long as you use these two functions instead of the original getters"
Are `msg.sender` and `msg.data` getters?

Reply:
Technically I guess they could be considered properties of a magical global `msg` struct. I dislike the 'getters' term though, do you have any suggestions?

Reply:
I don't have a better term than getters, I had a quick look through the solidity documentation but didn't see anything useful.

Perhaps just avoid using a term and specify them directly.

```suggestion
As long as you use these two replacement functions instead of `msg.sender` and `msg.data`, you're good to go!
```